kernel/kernel-api/src/main/java/io/delta/kernel/internal/skipping/DataSkippingUtils.java

```diff
@@ -0,0 +1,290 @@
+     *      AND(minValues.a + 1 <= 3, 3 <= maxValues.a + 1)
+     *
+     *    However, the skipping predicate would be unsound, because the addition operator could
+     *    trigger integer overflow (e.g. minValues.a = 0 and maxValues.a = INT_MAX), even though the
+     *    file could very well contain rows satisfying a + 1 = 3.
+     *
+     * 3. Predicates involving NOT are ineligible for skipping, because
+     *    `Not(constructDataFilters(e))` is seldom equivalent to `constructDataFilters(Not(e))`.
+     *    For example, consider the query predicate:
+     *
+     *      NOT(a = 1)
+     *
+     *    A simple inversion of the data skipping predicate would be:
+     *
+     *      NOT(AND(minValues.a <= 1, maxValues.a >= 1))
+     *      ==> OR(NOT(minValues.a <= 1), NOT(maxValues.a >= 1))
+     *      ==> OR(minValues.a > 1, maxValues.a < 1)
+     *
+     *    By contrast, if we first combine the NOT with = to obtain
+     *
+     *      a != 1
+     *
+     *    We get a different skipping predicate:
+     *
+     *      NOT(AND(minValues.a = 1, maxValues.a = 1))
+     *      ==> OR(NOT(minValues.a = 1), NOT(maxValues.a = 1))
+     *      ==>  OR(minValues.a != 1, maxValues.a != 1)
+     *
+     *    A truth table confirms that the first (naively inverted) skipping predicate is incorrect:
+     *
+     *      minValues.a
+     *      | maxValues.a
+     *      | | OR(minValues.a > 1, maxValues.a < 1)
+     *      | | | OR(minValues.a != 1, maxValues.a != 1)
+     *      0 0 T T
+     *      0 1 F T    !! first predicate wrongly skipped a = 0
+     *      1 1 F F
+     *
+     *    Fortunately, we may be able to eliminate NOT from some (branches of some) predicates:
+     *
+     *    a. It is safe to push the NOT into the children of AND and OR using de Morgan's Law, e.g.
+     *
+     *         NOT(AND(a, b)) ==> OR(NOT(a), NOT(B)).
+     *
+     *    b. It is safe to fold NOT into other operators, when a negated form of the operator
+     *       exists:
+     *
+     *         NOT(NOT(x)) ==> x
+     *         NOT(a == b) ==> a != b
+     *         NOT(a > b) ==> a <= b
+     *
+     * TODO update these docs about missing stats
+     * NOTE: The skipping predicate must handle the case where min and max stats for a column are
+     * both NULL -- which indicates that all values in the file are NULL. Fortunately, most of the
+     * operators we support data skipping for are NULL intolerant, and thus trivially satisfy this
+     * requirement because they never return TRUE for NULL inputs. The only NULL tolerant operator
+     * we support -- IS [NOT] NULL -- is specifically NULL aware.
+     *
+     * NOTE: The skipping predicate does *NOT* need to worry about missing stats columns (which also
+     * manifest as NULL). That case is handled separately by `verifyStatsForFilter` (which disables
+     * skipping for any file that lacks the needed stats columns).
+     */
+    public static Optional<Predicate> constructDataFilters(
```

Comment:
isnt it better call it constructDataSkippingFilters? data filters is confusing because these are not filters on the table data. in fact you take the dataFilter as a parameter, so the returned filter cannot be a dataFilter as well .. right?

Reply:
I copied names from delta-spark so the functions would match up... I'll go over and unify all the usages of terms instead